cloudprovider/aliyun/aliyun_provider.go

```diff
@@ -0,0 +1,91 @@
+package aliyun
+
+import (
+	"os"
+	"time"
+
+	"github.com/denverdino/aliyungo/metadata"
+	"github.com/rancher/agent/cloudprovider"
+	"github.com/rancher/agent/core/hostInfo"
+)
+
+const (
+	aliyunTag = "aliyun"
+)
+
+type Provider struct {
+	client      metadataClient
+	expireTime  time.Duration
+	interval    time.Duration
+	initialized bool
+}
+
+type metadataClient interface {
+	Region() (string, error)
+	Zone() (string, error)
+}
+
+type metadataClientImpl struct {
+	client *metadata.MetaData
+}
+
+func init() {
+	cloudprovider.AddCloudProvider(aliyunTag, &Provider{
+		expireTime: time.Minute * 5,
+		interval:   time.Second * 10,
+	})
+}
+
+func (m metadataClientImpl) Region() (string, error) {
+	return m.client.Region()
+}
+
+func (m metadataClientImpl) Zone() (string, error) {
+	return m.client.Zone()
+}
+
+func (p *Provider) Init() error {
+	client := metadataClientImpl{metadata.NewMetaData(nil)}
+	p.client = client
+	p.initialized = true
+	return nil
+}
+
+func (p *Provider) GetCloudProviderInfo() bool {
+	if !p.initialized {
+		return false
+	}
+	success := false
+	endtime := time.Now().Add(p.expireTime)
+	for {
+		if time.Now().After(endtime) {
+			break
+		}
+		if _, err := os.Stat(cloudprovider.InfoPath); err == nil {
```

Comment:
This process is exactly the same for both providers. We should able to abstract it. The interface should only contain how to get zone/region/tag.  Like this:

`func (p *Provider) GetLabels() (map[string]string{}, error)`

And the provider will retry when error.